pmd-core/src/main/java/net/sourceforge/pmd/internal/util/ClasspathClassLoader.java

```diff
@@ -53,6 +53,16 @@ public class ClasspathClassLoader extends URLClassLoader {
 
     static {
         registerAsParallelCapable();
+
+        // Disable caching for jar files to prevent issues like #4899
+        try {
+            // Uses a pseudo URL to be able to call URLConnection#setDefaultUseCaches
+            // with Java9+ there is a static method for that per protocol:
+            // https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/net/URLConnection.html#setDefaultUseCaches(java.lang.String,boolean)
+            URI.create("jar:file:file.jar!/").toURL().openConnection().setDefaultUseCaches(false);
```

Comment:
I find it a little disturbing that there is very little documentation as to what `openConnection()` does or when it may throw an `IOException`. For HTTP/HTTPS it's explicitly stated that no connection is open until `connect()` is invoked, but for files I've found nothing at all…

We are effectively relying on `openConnection()` not checking or touching the file for this to work…